samples/java-spring-choreography-saga-quickstart/src/main/java/user/registry/api/ApplicationController.java

```diff
@@ -0,0 +1,190 @@
+
+import java.util.Optional;
+
+/**
+ * Controller for the user registry application.
+ * This controller works as a gateway for the user service. It receives the requests from the outside world and
+ * forwards them to the user service and ensure that the email address is not already reserved.
+ * <p>
+ * The UniqueEmailEntity and the UserEntity are protected from external access. They can only be accessed through
+ * this controller.
+ */
+@RequestMapping("/api")
+public class ApplicationController extends Action {
+
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+  private final ComponentClient client;
+
+  public ApplicationController(ComponentClient client) {
+    this.client = client;
+  }
+
+
+  /**
+   * External API representation of an email record
+   */
+  public record EmailInfo(String address, String status, Optional<String> ownerId) {
+  }
+
+  /**
+   * External API representation of a User
+   */
+  public record UserInfo(String id, String name, String country, String email) {
+  }
+
+  /**
+   * This is the main entry point for creating a new user.
+   * <p>
+   * Before creating a User, we need to reserve the email address to ensure that it is not already used.
+   * The call will fail if the email address is already reserved.
+   * <p>
+   * If we succeed in reserving the email address, we move forward and create the user.
+   */
+  @PostMapping("/users/{userId}")
+  public Effect<Done> createUser(@PathVariable String userId, @RequestBody UserEntity.Create cmd) {
+
+    var createUniqueEmail = new UniqueEmailEntity.ReserveEmail(cmd.email(), userId);
+
+    logger.info("Reserving new address '{}'", cmd.email());
+    // eagerly, reserving the email address
+    // we want to execute this call in other to check its result
+    // and decide if we can continue with the user creation
+    var emailReserved =
+      client
+        .forValueEntity(cmd.email())
+        .call(UniqueEmailEntity::reserve)
+        .params(createUniqueEmail)
+        .execute(); // eager, executing it now
+
+    // this call is lazy and will be executed only if the email reservation succeeds
+    var callToUser =
+      client
+        .forEventSourcedEntity(userId)
+        .call(UserEntity::createUser)
+        .params(cmd);
+
+
+    var userCreated =
+      emailReserved
+        .thenApply(__ -> {
+          // on successful email reservation, we create the user and return the result
+          logger.info("Creating user '{}'", userId);
+          return effects().asyncReply(callToUser.execute());
+        })
+        .exceptionally(e -> {
+          // in case of exception `callToUser` is not executed,
+          // and we return an error to the caller of this method
+          logger.info("Email is already reserved '{}'", cmd.email());
+          return effects().error("Email is already reserved '" + cmd.email() + "'", StatusCode.ErrorCode.BAD_REQUEST);
+        });
+
+    return effects().asyncEffect(userCreated);
+  }
```

Comment:
I think this is inherent to the nature of a system relying on async calls. The solution here is to use Java 21 and have some alternative APIs. 

We would need some alternative API, because if we remove the CompletionStage, we will still need something to branch between successful execution and failure.

A Java 21 friendly API could be something like:

```java
effect
  .replyWith(deferredCall) // execute call, get results and build Effect[R]
  .onError(throwable -> throwable.getMessage) // from throwable to effect.error
```

That with the extra challenge that we such API should only be used with Java 21, otherwise it will be pure old blocking. 🤷